### DISCUSSION TITLE
Problem with SET search_path in Postgres

### DISCUSSION THREAD

**Discussion by Nahuelsegovia** [sqlalchemy/sqlalchemy#11289] [Usage Questions]
I have been working with sqlAlchemy and I have a problem when I try to execute SET search_path in Postgres.

Postgres with Postgis docker-compose.yml

```bash
version: '3.9'
services:
  postgres:
    image: postgis/postgis
    container_name: postgres
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: posgres
    ports:
      - "5432:5432"
 ```

get_sa_session function

```python

def get_sa_session():
    session = None
    engine = create_engine(
        'postgresql://' +
        os.environ.get("DB_USER")+':'+os.environ.get("DB_PASSWORD") +
        '@'+os.environ.get("DB_HOST")+':'+str(os.environ.get("DB_PORT")) +
        '/'+os.environ.get("DB_DATABASE"),
        poolclass=NullPool
                     
    )
    session = sessionmaker(bind=engine)()
    return session

```

POC

```python

    def test1(self):
        try:
            session = get_sa_session()
            session.execute(text("SET search_path TO unit_test_schema"))
            session.commit()
            session.flush()


            result = session.execute(text("SHOW search_path"))

            search_path = result.fetchone()[0]

            print("Current search path: ", search_path)
```

![Screenshot from 2024-04-18 08-56-52](https://github.com/sqlalchemy/sqlalchemy/assets/13077988/334c6cda-90b4-47cb-b19a-5d57d8038f94)

Executing SET search_path TO unit_test_schema in DBeaver

![Screenshot from 2024-04-18 08-59-07](https://github.com/sqlalchemy/sqlalchemy/assets/13077988/b9e0a26e-f812-4d00-8c4b-0fd3a5b6c129)

O.S
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.4 LTS
Release:	22.04
Codename:	jammy

SQLalchemy version
2.0.23

Python version
Python 3.11.0rc1

**Comment by zzzeek** [comment]
you are setting the search path on only one PostgreSQL connection but not others that may be pulled from the connection pool.  When you call `session.commit()`, the connection is released to the pool; on next use, since you are using `NullPool` you are guaranteed to get a new connection that doesnt have this search path set.

solution, ensure your search path setting is engine-wide for all connections unconditionally, and only use the engine in that setting.  See https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#setting-alternate-search-paths-on-connect for how to do this.

**Reply by Nahuelsegovia** [reply]
Thanks! I could connect with another schema successfully :)
